File: apps/web/lib/inbox.ts
/**
 * Client-side research inbox for the Go API (T1003).
 *
 * GET  /api/v1/inbox            one page of aggregated notification entries
 * POST /api/v1/inbox/read       mark the entries the caller read
 * POST /api/v1/inbox/read-all   mark what the inbox is serving read
 *
 * An ENTRY is not a notification: it is the deliveries of one target and
 * one event type inside one clock-hour window, rendered as one row with a
 * count (internal/events/inbox.go). That is what keeps a project's
 * per-commit traffic from becoming a list of forty rows — the count is
 * how the list says "and 39 more" without hiding any of them. The page
 * never re-derives the aggregation: `count`, `unread` and the window come
 * from the API, which is the one place that knows how entries are formed.
 *
 * Read state travels the same way. `latest_delivery_id` is the newest
 * delivery the page was shown, and marking read names it: the API marks
 * that entry up to that delivery, so a notification that arrived after
 * the page was rendered stays unread. The client therefore marks and then
 * re-reads — recomputing the badge locally would be a second
 * implementation of a rule the API already owns.
 *
 * Every call carries the session cookie (credentials: include); the two
 * writes also head the session-bound CSRF token (docs/22), which the
 * caller supplies from lib/auth's sessionTokenStorage.
 *
 * The module is import-free by construction (like lib/milestones.ts), so
 * the node:test suite (inbox.test.mjs) runs it through Node's type
 * stripping. The ApiError shape mirrors lib/projects.ts.
 */

/** The wire entry shape (cmd/api/inboxhttp inboxEntryPayload). */
export interface InboxEntry {
  /** project | asset | user | organization | knowledge. */
  target_type: string;
  target_id: string;
  /**
   * The target's display name, resolved by the API when the inbox is read.
   *
   * "" only when the API has no naming query for that target KIND yet (see
   * entryTargetLabel for what to render instead). It never means "you may
   * not read this target": an entry whose target the caller may no longer
   * read is not in the page at all — no name, no id, no count, not in the
   * badge — so a client that reads "" as "no access" would state something
   * about a row it was never sent.
   */
  target_label: string;
  /** The canonical event type (specs/events/event-types.yaml). */
  event_type: string;
  /** RFC 3339 — the start of the clock-hour window the entry aggregates. */
  window_start: string;
  /** How many deliveries this one entry stands for. */
  count: number;
  /** How many of them are unread. */
  unread: number;
  /** unread === 0, as the API computes it. */
  read: boolean;
  first_at: string;
  last_at: string;
  latest_event_id: string;
  /** The anchor a mark-read call names (the newest delivery shown). */
  latest_delivery_id: string;
  /** The subject's route, or "" for a target with no page yet. */
  url: string;
}

/** One page of the inbox (GET /api/v1/inbox). */
export interface InboxPage {
  entries: InboxEntry[];
  /** The badge: entries holding anything unread, over the whole inbox. */
  unreadCount: number;
  /** The view that was served: "unread" or "all". */
  filter: string;
}

/** The two views the API serves (internal/events/inbox.go). */
export const INBOX_FILTER_UNREAD = "unread";
export const INBOX_FILTER_ALL = "all";

/** The API error envelope (docs/22 §5): stable codes, no stack traces. */
export interface ErrorEnvelope {
  code: string;
  message: string;
  request_id?: string;
  retryable?: boolean;
}

/**
 * ApiError carries the API's envelope: the UI renders `message` for the
 * codes it knows and a generic line otherwise — never raw transport
 * errors. (Same shape as lib/projects.ts ApiError.)
 */
export class ApiError extends Error {
  readonly code: string;
  readonly status: number;
  readonly retryable: boolean;

  constructor(status: number, envelope: ErrorEnvelope) {
    super(envelope.message);
    this.name = "ApiError";
    this.code = envelope.code;
    this.status = status;
    this.retryable = envelope.retryable ?? status === 503;
  }
}

/** The inbox client: one per API origin. */
export interface InboxClient {
  /**
   * One page of the caller's inbox, newest entry first. filter defaults
   * to the unread view; limit is the page size the API clamps.
   */
  list(filter?: string, limit?: number): Promise<InboxPage>;
  /**
   * Mark the named entries read. The ids are the entries'
   * latest_delivery_id values — the anchors the API marks up to.
   * Returns how many deliveries were marked; an empty list is a no-op
   * that calls nothing.
   */
  markRead(deliveryIds: string[]): Promise<number>;
  /**
   * Mark what the inbox is serving read — every unread entry the API
   * would show the caller, and nothing else: an entry whose target the
   * caller may no longer read is not marked, so it stays unread and comes
   * back unread if access returns. Returns how many deliveries were
   * marked, so a button's "cleared" line counts what was really cleared.
   */
  markAllRead(): Promise<number>;
}

type FetchLike = (
  input: string,
  init?: {
    method?: string;
    headers?: Record<string, string>;
    body?: string;
    credentials?: string;
  },
) => Promise<{ status: number; json(): Promise<unknown> }>;

/** Options for createInboxClient. */
export interface InboxClientOptions {
  /** Fetch implementation (tests inject a fake); default: global fetch. */
  fetch?: FetchLike;
  /**
   * The session-bound CSRF token for the two writes (the API's guard
   * demands it on every state change). The module stays import-free, so
   * the caller supplies the source — the page reads lib/auth's
   * sessionTokenStorage.
   */
  csrfToken?: () => string | null;
}

const CSRF_HEADER = "X-CSRF-Token";

/** Build the inbox client for one API origin. */
export function createInboxClient(
  apiBaseUrl: string,
  options: InboxClientOptions = {},
): InboxClient {
  const fetchFn: FetchLike = options.fetch ?? (fetch as unknown as FetchLike);

  /** One JSON call against the API; the CSRF token (when known) heads every call. */
  async function call(
    method: string,
    path: string,
    body?: unknown,
  ): Promise<{ status: number; body: unknown }> {
    const headers: Record<string, string> = {};
    const csrf = options.csrfToken?.() ?? null;
    if (csrf !== null) headers[CSRF_HEADER] = csrf;
    if (body !== undefined) headers["Content-Type"] = "application/json";
    const res = await fetchFn(apiBaseUrl + path, {
      method,
      headers,
      body: body === undefined ? undefined : JSON.stringify(body),
      credentials: "include",
    });
    let parsed: unknown = null;
    try {
      parsed = await res.json();
    } catch {
      parsed = null; // non-JSON bodies (e.g. a proxy error page)
    }
    return { status: res.status, body: parsed };
  }

  /** Throw ApiError for any non-2xx, decoding the stable envelope. */
  function fail(status: number, body: unknown): never {
    if (typeof body === "object" && body !== null && "code" in body) {
      throw new ApiError(status, body as ErrorEnvelope);
    }
    throw new ApiError(status, {
      code: "UNKNOWN",
      message: "the server answered unexpectedly; try again",
    });
  }

  /** One entry, shape-checked: a page that renders half an entry lies. */
  function asEntry(body: unknown): InboxEntry {
    if (
      typeof body !== "object" || body === null ||
      !("target_type" in body) || typeof body.target_type !== "string" ||
      !("target_id" in body) || typeof body.target_id !== "string" ||
      !("event_type" in body) || typeof body.event_type !== "string" ||
      !("window_start" in body) || typeof body.window_start !== "string" ||
      !("count" in body) || typeof body.count !== "number" ||
      !("unread" in body) || typeof body.unread !== "number" ||
      !("latest_delivery_id" in body) || typeof body.latest_delivery_id !== "string"
    ) {
      throw new Error("unexpected inbox entry shape");
    }
    return body as InboxEntry;
  }

  /** The read count the two write routes answer with. */
  async function markedCount(
    path: string,
    body: unknown | undefined,
  ): Promise<number> {
    const result = await call("POST", path, body);
    if (result.status >= 400) fail(result.status, result.body);
    const payload = result.body;
    if (
      typeof payload !== "object" || payload === null ||
      !("read" in payload) || typeof payload.read !== "number"
    ) {
      throw new Error("unexpected mark-read response shape");
    }
    return payload.read;
  }

  return {
    async list(filter = INBOX_FILTER_UNREAD, limit) {
      const params = new URLSearchParams({ filter });
      if (limit !== undefined) params.set("limit", String(limit));
      const result = await call("GET", `/api/v1/inbox?${params.toString()}`);
      if (result.status >= 400) fail(result.status, result.body);
      const body = result.body;
      if (
        typeof body !== "object" || body === null ||
        !("entries" in body) || !Array.isArray(body.entries) ||
        !("unread_count" in body) || typeof body.unread_count !== "number" ||
        !("filter" in body) || typeof body.filter !== "string"
      ) {
        throw new Error("unexpected inbox response shape");
      }
      return {
        entries: body.entries.map(asEntry),
        unreadCount: body.unread_count,
        filter: body.filter,
      };
    },
    async markRead(deliveryIds) {
      // An empty selection is what the API would answer 0 to; calling it
      // would be a request that can only say what the caller already
      // knows, and the caller's UI should not offer the action at all.
      if (deliveryIds.length === 0) return 0;
      return markedCount("/api/v1/inbox/read", { deliveries: deliveryIds });
    },
    async markAllRead() {
      return markedCount("/api/v1/inbox/read-all", undefined);
    },
  };
}

/** Human-facing message for a stable inbox API code. */
export function messageForInboxCode(code: string): string {
  switch (code) {
    case "AUTH_UNAUTHENTICATED":
      return "Sign in to read your inbox.";
    case "INBOX_ENTRY_NOT_FOUND":
      return "That notification is no longer in your inbox. Reload the page.";
    case "VALIDATION_FAILED":
      return "The inbox request was rejected. Reload the page and try again.";
    case "SERVICE_UNAVAILABLE":
      return "Inbox data is temporarily unavailable. Please try again later.";
    default:
      return "Something went wrong while reading the inbox. Please try again.";
  }
}

/**
 * The display name of one canonical event type (specs/events/event-types.yaml).
 * An unknown type renders RAW — the event's own name is the truth, and a
 * made-up label for it would be the page inventing product semantics.
 */
const EVENT_TYPE_NAMES: Record<string, string> = {
  "project.created": "Project created",
  "project.visibility_changed": "Project visibility changed",
  "project.main_frozen": "Main frozen",
  "branch.created": "Branch created",
  "branch.aborted": "Branch aborted",
  "state.committed": "Commit",
  "pull_request.opened": "Pull request opened",
  "pull_request.reviewed": "Pull request reviewed",
  "pull_request.merged": "Pull request merged",
  "scientific_object.version_created": "Object version created",
  "scientific_object.aborted": "Object aborted",
  "scientific_object.reopened": "Object reopened",
  "evidence_assertion.created": "Evidence assertion recorded",
  "release.published": "Release published",
  "research_asset.version_published": "Asset version published",
  "knowledge.version_published": "Knowledge published",
  "knowledge.external_evidence_added": "External evidence added",
  "dependency.status_changed": "Dependency status changed",
  "dependency.impact_detected": "Dependency impact detected",
  "contribution.accepted": "Contribution accepted",
  "credit.dispute_opened": "Credit dispute opened",
  "credit.dispute_resolved": "Credit dispute resolved",
  "rights.visibility_changed": "Rights visibility changed",
  "policy.version_published": "Policy published",
  "webhook.delivery_failed": "Webhook delivery failed",
};

/** The display name of one event type; unknown types render raw. */
export function eventTypeDisplayName(eventType: string): string {
  return EVENT_TYPE_NAMES[eventType] ?? eventType;
}

/**
 * The aggregation badge of one entry: how many notifications it stands
 * for, or null when it stands for one.
 *
 * The threshold is 2 because that is where the aggregation becomes
 * meaningful: a row that says "1 event" beside every notification reads
 * as noise, while a row that says "40 events" is the whole reason the
 * inbox exists in this shape. The number is rendered as the API counted
 * it — never recomputed here.
 */
export function aggregationLabel(count: number): string | null {
  if (!Number.isFinite(count) || count < 2) return null;
  return `${count} events`;
}

/**
 * What a row calls its subject.
 *
 * target_label is "" only when the API has no naming query for that target
 * KIND (knowledge, until it has a surface) — never because the caller may
 * not read the target: such an entry is not in the page at all, so this
 * fallback is reached for rows the caller IS being served. A row that
 * rendered nothing there would look broken, so the fallback names the
 * target's kind and a short id.
 *
 * Do not render an "access denied" state on "" — the row would be claiming
 * the caller was shown something they cannot see, which the API cannot do.
 */
export function entryTargetLabel(entry: InboxEntry): string {
  if (entry.target_label !== "") return entry.target_label;
  const short = entry.target_id.length > 8 ? entry.target_id.slice(0, 8) : entry.target_id;
  return `${entry.target_type} ${short}`;
}

/**
 * One entry's window as a row renders it: the clock hour, UTC.
 *
 * The window is a UTC hour by construction — the API bins deliveries with
 * date_bin from the epoch — so it is labelled UTC rather than silently
 * shifted into the viewer's timezone: two rows claiming different hours
 * for the same deliveries would be one of them lying. An unparseable
 * timestamp renders "" (the row shows no window rather than "Invalid
 * Date").
 */
export function windowLabel(windowStart: string): string {
  const at = new Date(windowStart);
  if (Number.isNaN(at.getTime())) return "";
  const start = at.toISOString().slice(0, 13).replace("T", " ") + ":00";
  const end = new Date(at.getTime() + 3_600_000).toISOString().slice(11, 13);
  return `${start}–${end}:00 UTC`;
}

File: internal/events/inbox.go
package events

import (
	"fmt"
	"time"
)

// The research inbox (T1003, docs/18 §3-4): the WEB channel's read model
// over the delivery rows the subscription fan-out writes (T1002).
//
// A delivery row is the notification — T1002 made exactly one row per
// (subscription, event, channel) and called the web channel's row
// 'delivered' the moment it exists, "the inbox row IS the delivery". What
// this file adds is the two things the rows alone do not answer:
//
//  1. AGGREGATION. "默认避免通知每个低级 commit" (docs/18 §3) is a rule
//     about granularity, not about dropping events: a commit is worth
//     knowing about, and fifty commits in one hour are worth knowing about
//     ONCE. An inbox ENTRY is therefore the deliveries of one target and
//     one event type inside one fixed window of InboxAggregationWindow,
//     rendered as one row with a count — not fifty rows, and not a
//     summarization that loses the identity of what happened. The entry
//     keeps the newest delivery's event id so the surface can still link
//     to the event itself, and the count so it can say "and 49 more".
//
//     The window is fixed clock time, not a gap between deliveries, and
//     that is deliberate: entry membership has to be a pure function of
//     the delivery rows, because the read is a GROUP BY over them rather
//     than a stored object (see migration 00079 for why read state lives
//     on the delivery). A gap-based rule would make an entry's boundaries
//     depend on the order rows were inserted, so a delivery that arrived
//     late — or one the revocation path cancelled — could silently move
//     earlier deliveries into a different entry than the one a subscriber
//     already read.
//
//  2. READ/UNREAD. read_at on the delivery row; an entry is unread while
//     any of its deliveries is. Marking read is scoped to what was SHOWN:
//     the caller names the delivery it read and everything older in that
//     same entry is marked, so a notification that arrives a second later
//     is not swept into a read it never had.
//
//  3. AUDIENCE. The read is fail-closed on the caller's CURRENT relation
//     to each target: a target that resolves to AudienceNone (a project
//     they were removed from and made private, a deactivated account)
//     takes its entries out of the inbox entirely — out of both views and
//     out of the badge. The audience is resolved over the WHOLE inbox
//     before any page is cut, so the row a caller sees and the number
//     they see are computed from the same set (inboxVisibleTargets). A
//     delivery row is not deleted by this and not rewritten: it is the
//     caller's own record that something was delivered, and the rule is
//     about what may still be rendered from it.
//
//     The WRITE obeys the same rule, in the same resolution: read-all
//     marks exactly the rows the read would serve (MarkInboxAllRead), so a
//     row the caller cannot see is never marked read on their behalf. What
//     a caller may be shown and what may be marked read are two halves of
//     one question, and answering them from one inboxVisibleTargets is
//     what keeps the badge, the page and the button from disagreeing.

// InboxAggregationWindow is how much wall-clock time one inbox entry
// spans: deliveries of the same (target, event type) inside one window of
// this size are ONE entry.
//
// One hour is the L1 choice, and it is a compromise between the two ways
// an inbox goes wrong. Too small and a working session's worth of commits
// is still a list (the bombardment docs/18 §3 forbids); too large and
// genuinely separate pieces of work — a morning's commits and an
// afternoon's — arrive as one row that cannot be told apart. Within the
// window nothing is lost: the entry carries its delivery count, the
// window's start and end, and the newest event's identity.
const InboxAggregationWindow = time.Hour

// The inbox views. "unread" is the default: the inbox is a to-do list
// first, and the rows already read are history it can still show on
// request.
const (
	// InboxFilterUnread lists entries with at least one unread delivery.
	InboxFilterUnread = "unread"
	// InboxFilterAll lists every entry, read ones included.
	InboxFilterAll = "all"
)

// Inbox read bounds. The default is what one page renders; the maximum is
// what a caller may ask for. A limit is not a horizon — nothing is
// excluded from the read by it, only from the page.
const (
	DefaultInboxLimit = 50
	MaxInboxLimit     = 200
)

// ErrInboxFilter: the view name is not one of the inbox's.
var ErrInboxFilter = fmt.Errorf("events: invalid inbox filter")

// ErrInboxDeliveryNotFound: no delivered web delivery with that id belongs
// to the caller. "not yours" and "does not exist" answer identically, the
// same existence-hiding rule the subscription reads keep.
var ErrInboxDeliveryNotFound = fmt.Errorf("events: inbox delivery not found")

// InboxEntry is one aggregated inbox row: the deliveries of one user, one
// target and one event type inside one window (see the file comment).
//
// Its identity is the tuple (TargetType, TargetID, EventType,
// WindowStart) — that is what a caller is looking at, and it is stable
// across reads, because it is derived from the deliveries rather than
// assigned. LatestDeliveryID is the newest delivery IN the entry and is
// the anchor a mark-read call names; it moves as new deliveries arrive,
// which is the point: reading marks what was shown.
type InboxEntry struct {
	TargetType string
	TargetID   string
	// TargetLabel is the target's display identity — a project name, an
	// asset title, a person's handle, an organization's name — read at
	// render time rather than copied into the delivery when it was fanned
	// out. It is "" only for a target type that has no naming query yet
	// (see inboxLabelQueries); a target the caller may no longer read
	// produces no entry at all, so its name never reaches this struct.
	TargetLabel string
	EventType   string
	WindowStart time.Time
	// Count is how many deliveries the entry aggregates.
	Count int
	// Unread is how many of them are unread. Every delivery counted here
	// is a delivered web row (status 'delivered') whose target the caller
	// can still read: a cancelled row is not in the entry at all, and a
	// target whose audience no longer includes the caller takes its whole
	// entry out of the inbox rather than shrinking it
	// (inboxVisibleTargets, decision 3).
	Unread  int
	FirstAt time.Time
	LastAt  time.Time
	// LatestEventID is the research event of the newest delivery — the
	// identity a surface links to the event itself with.
	LatestEventID string
	// LatestDeliveryID is the newest delivery row, and the anchor for
	// marking this entry read.
	LatestDeliveryID string
}

// Read reports whether the entry has been read. An entry with no unread
// deliveries is read — including an entry that was read and then grew no
// further.
func (e InboxEntry) Read() bool { return e.Unread == 0 }

// Target renders the entry's aggregation key minus the window.
func (e InboxEntry) Target() Target {
	return Target{Type: e.TargetType, ID: e.TargetID}
}

// ValidateInboxFilter checks an inbox view name at the boundary, so an
// unknown view answers a validation error rather than silently serving a
// different list than the caller asked for.
func ValidateInboxFilter(filter string) error {
	switch filter {
	case InboxFilterUnread, InboxFilterAll:
		return nil
	}
	return fmt.Errorf("%w: unknown inbox filter %q (want %q or %q)",
		ErrInboxFilter, filter, InboxFilterUnread, InboxFilterAll)
}

// InboxLimit clamps a requested page size into the legal range: a
// non-positive request means the default (a caller that names no limit
// wants a page, not zero rows), and anything above the maximum is capped
// rather than refused, because the cap is the server's protection and not
// a client error.
func InboxLimit(requested int) int {
	if requested <= 0 {
		return DefaultInboxLimit
	}
	if requested > MaxInboxLimit {
		return MaxInboxLimit
	}
	return requested
}
